Rework StreamBuilder to avoid unnecessary transcoding

eyezak opened the pull request:
**Changes**
Adds preferred matching of video and audio streams to both DirectPlay and Transcode profiles. Multiple candidate profiles are considered after codec resolution, with preference to copy-codec. In the case of compatible video streams, remuxing is preferred to transcoding audio. In the cases of no compatible video streams,  multiple equivalent candidate profiles and !Allow*Copy - preference is given to the client-supplied profile order to support prefer fMP4/ts streams user-setting. In contrast to the approach of https://github.com/jellyfin/jellyfin/pull/6454, PlaybackInfo based apis now prefer transcoding to direct streaming to allow direct-stream equivalent transcoding to HLS (ts or fMP4).
MediaInfoHelper has been adapted to only call StreamBuilder once.

Converts TranscodeReason to Flags and serialize into strings in models for simplicity.
Adds a suite of tests for StreamBuilder based on real and imagined device profiles.

Currently testing on Firefox/Chrome/Safari with av1,divx,h264,hevc,msmp4,xvid x aac,ac3,eac3,mp3,flac,opus,vorbis x mp4,mkv,avi,webm encodes of the Sintel 2048 trailer. Verified HEVC10 and HEVC on safari with remux and transcode as direct-stream with web-player modifications (https://github.com/jellyfin/jellyfin-web/pull/3442) .

**Issues**
fixes #6450
fixes #6149
alternative approach to #6454 
Potentially others i have not validated yet

crobibero commented on MediaBrowser.Model/Dlna/StreamBuilder.cs:
```diff
@@ -15,6 +15,12 @@ namespace MediaBrowser.Model.Dlna
     public class StreamBuilder
     {
+        // Aliases
+        public static readonly TranscodeReason ContainerReasons = TranscodeReason.ContainerNotSupported | TranscodeReason.ContainerBitrateExceedsLimit;
+        public static readonly TranscodeReason AudioReasons = TranscodeReason.AudioCodecNotSupported | TranscodeReason.AudioBitrateNotSupported | TranscodeReason.AudioChannelsNotSupported | TranscodeReason.AudioProfileNotSupported | TranscodeReason.AudioSampleRateNotSupported | TranscodeReason.SecondaryAudioNotSupported | TranscodeReason.AudioBitDepthNotSupported | TranscodeReason.AudioIsExternal ...
+        public static readonly TranscodeReason VideoReasons = TranscodeReason.VideoCodecNotSupported | TranscodeReason.VideoResolutionNotSupported | TranscodeReason.AnamorphicVideoNotSupported | TranscodeReason.InterlacedVideoNotSupported | TranscodeReason.VideoBitDepthNotSupported | TranscodeReason.VideoBitrateNotSupported | TranscodeReason.VideoFramerateNotSupported | TranscodeReason.VideoLevel ...
+        private static readonly TranscodeReason DirectStreamReasons = AudioReasons | TranscodeReason.ContainerNotSupported;
```
```suggestion
        internal const TranscodeReason ContainerReasons = TranscodeReason.ContainerNotSupported | TranscodeReason.ContainerBitrateExceedsLimit;
        internal const TranscodeReason AudioReasons = TranscodeReason.AudioCodecNotSupported | TranscodeReason.AudioBitrateNotSupported | TranscodeReason.AudioChannelsNotSupported | TranscodeReason.AudioProfileNotSupported | TranscodeReason.AudioSampleRateNotSupported | TranscodeReason.SecondaryAudioNotSupported | TranscodeReason.AudioBitDepthNotSupported | TranscodeReason.AudioIsExternal;
        internal const TranscodeReason VideoReasons = TranscodeReason.VideoCodecNotSupported | TranscodeReason.VideoResolutionNotSupported | TranscodeReason.AnamorphicVideoNotSupported | TranscodeReason.InterlacedVideoNotSupported | TranscodeReason.VideoBitDepthNotSupported | TranscodeReason.VideoBitrateNotSupported | TranscodeReason.VideoFramerateNotSupported | TranscodeReason.VideoLevelNotSupported | TranscodeReason.RefFramesNotSupported;
        internal const TranscodeReason DirectStreamReasons = AudioReasons | TranscodeReason.ContainerNotSupported;
```

And add the following to `MediaBrowser.Model/Properties/AssemblyInfo.cs` (under `[assembly: InternalsVisibleTo("Jellyfin.Model.Tests")]`

```c#
[assembly: InternalsVisibleTo("Jellyfin.Dlna.Tests")]
```

crobibero commented on MediaBrowser.Model/Session/TranscodeReason.cs:
```diff
@@ -1,32 +1,46 @@
+
 namespace MediaBrowser.Model.Session
 {
+    [Flags]
```
Just as a note- changing this enum to `Flags` changed the integer representation of each value. This shouldn't matter as the API returns the string.

eyezak replied:
yeah, i saw that the api was string based and I chose to change the values and have the sort-order be meaningful as well

eyezak commented:
I added another commit. I found during rebase I missed a transcode-matching case - added tests and updated SafariNext test device profile to match https://github.com/jellyfin/jellyfin-web/pull/3442

crobibero commented on MediaBrowser.Model/Dlna/StreamBuilder.cs:
```diff
@@ -689,189 +628,299 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
-                }
+                    var videoCodecs = ContainerProfile.SplitValue(transcodingProfile.VideoCodec);
+
+                    if (ContainerProfile.ContainsContainer(videoCodecs, item.VideoStream.Codec)) {
```
```suggestion
                    if (ContainerProfile.ContainsContainer(videoCodecs, item.VideoStream.Codec))
                    {
```

Bond-009 commented on MediaBrowser.Model/Session/TranscodeReasonExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// Extension methods for serializing TranscodeReason.
+    /// </summary>
+    public static class TranscodeReasonExtensions
+    {
+        private static readonly TranscodeReason[] _values = Enum.GetValues<TranscodeReason>();
+
+        /// <summary>
+        /// Serializes a TranscodeReason into a delimiter-separated string.
+        /// </summary>
+        /// <param name="reasons">The <see cref="TranscodeReason"/> enumeration.</param>
+        /// <param name="sep">The string separator to use. defualt <c>,</c>.</param>
+        /// <returns>string of transcode reasons delimited.</returns>
+        public static string Serialize(this TranscodeReason reasons, string sep = ",")
+        {
+            return string.Join(sep, reasons.ToArray());
+        }
+
+        /// <summary>
+        /// Serializes a TranscodeReason into an array of individual TranscodeReason bits.
+        /// </summary>
+        /// <param name="reasons">The <see cref="TranscodeReason"/> enumeration.</param>
+        /// <returns>Array of <c>TranscodeReason</c>.</returns>
+        public static TranscodeReason[] ToArray(this TranscodeReason reasons)
+        {
+            return _values.Where(r => r != 0 && reasons.HasFlag(r)).ToArray();
+        }
+    }
```
`TranscodeReason.ToString()` should do exactly this afaik?

Bond-009 commented on MediaBrowser.Model/Dlna/StreamBuilder.cs:
```diff
@@ -689,189 +628,300 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
-            playlistItem.TranscodeReasons = transcodeReasons.ToArray();
+                // Make sure the video bitrate is lower than bitrate settings but at least 64k
+                long currentValue = playlistItem.VideoBitrate ?? availableBitrateForVideo;
+                var longBitrate = Math.Max(Math.Min(availableBitrateForVideo, currentValue), 64000);
```
`Math.Clamp` exists

Bond-009 commented on MediaBrowser.Model/Dlna/StreamBuilder.cs:
```diff
@@ -689,189 +628,300 @@ private StreamInfo BuildVideoItem(MediaSourceInfo item, VideoOptions options)
+                // Make sure the video bitrate is lower than bitrate settings but at least 64k
+                long currentValue = playlistItem.VideoBitrate ?? availableBitrateForVideo;
+                var longBitrate = Math.Max(Math.Min(availableBitrateForVideo, currentValue), 64000);
+                playlistItem.VideoBitrate = longBitrate >= int.MaxValue ? int.MaxValue : Convert.ToInt32(longBitrate);
```
Not sure about the `Convert.ToInt32`, only thing it does it trow if `longBitrate < int.MinValue`

Bond-009 commented on src/Jellyfin.Extensions/Json/JsonDefaults.cs:
```diff
@@ -36,6 +36,7 @@ public static class JsonDefaults
                 new JsonGuidConverter(),
                 new JsonNullableGuidConverter(),
                 new JsonVersionConverter(),
+                new JsonFlagEnumConverterFactory(),
```
Does this work as expected? When is `JsonStringEnumConverter` used now?

crobibero replied:
`JsonStringEnumConverter` is used for any enum that is not `[Flags]`

crobibero commented on Jellyfin.Api/Controllers/UniversalAudioController.cs:
```diff
@@ -223,7 +224,7 @@ public class UniversalAudioController : BaseJellyfinApiController
                     RequireNonAnamorphic = false,
                     EnableMpegtsM2TsMode = false,
-                    TranscodeReasons = mediaSource.TranscodeReasons == null ? null : string.Join(',', mediaSource.TranscodeReasons.Select(i => i.ToString()).ToArray()),
+                    TranscodeReasons = mediaSource.TranscodeReasons == 0 ? null : mediaSource.TranscodeReasons.ToString(),
```
Originally did have `None`, but the JsonConverter and ToString would always list None as a value....

Bond-009 replied:
That's stupid :/

crobibero commented on MediaBrowser.Controller/MediaEncoding/EncodingJobInfo.cs:
```diff
@@ -34,25 +35,23 @@ public EncodingJobInfo(TranscodingJobType jobType)
         }
 
-        public TranscodeReason[] TranscodeReasons
+        public TranscodeReason TranscodeReason
```
That's an API breaking change

nyanmisaka commented:
```
[16:41:17] [ERR] [41] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL POST /Items/f904cec75fb51710ef3db98195be9787/PlaybackInfo.
System.ArgumentException: '14616000' cannot be greater than 64000.
   at System.Math.ThrowMinMaxException[T](T min, T max)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildStreamVideoItem(StreamInfo playlistItem, VideoOptions options, MediaSourceInfo item, MediaStream videoStream, MediaStream audioStream, IEnumerable`1 candidateAudioStreams, String container, String videoCodec, String audioCodec)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(MediaSourceInfo item, VideoOptions options)
   at MediaBrowser.Model.Dlna.StreamBuilder.BuildVideoItem(VideoOptions options)
   at Jellyfin.Api.Helpers.MediaInfoHelper.SetDeviceSpecificData(BaseItem item, MediaSourceInfo mediaSource, DeviceProfile profile, AuthorizationInfo auth, Nullable`1 maxBitrate, Int64 startTimeTicks, String mediaSourceId, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String playSessionId, Guid userId, Boolean enableDirectPlay, Boolean enableDirectStream, Boolean enableTranscoding, Boolean allowVideoStreamCopy, Boolean allowAudioStreamCopy, IPAddress ipAddress)
   at Jellyfin.Api.Controllers.MediaInfoController.GetPostedPlaybackInfo(Guid itemId, Nullable`1 userId, Nullable`1 maxStreamingBitrate, Nullable`1 startTimeTicks, Nullable`1 audioStreamIndex, Nullable`1 subtitleStreamIndex, Nullable`1 maxAudioChannels, String mediaSourceId, String liveStreamId, Nullable`1 autoOpenLiveStream, Nullable`1 enableDirectPlay, Nullable`1 enableDirectStream, Nullable`1 enableTranscoding, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudioStreamCopy, PlaybackInfoDto playbackInfoDto)
   at lambda_method1192(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
```

nyanmisaka commented:
@eyezak Hi! I found that after this PR, media can be played directly on Chrome regardless of the resolution and bitrate limitations set by jellyfin-web. Any ideas on this?

dmitrylyzo commented:
https://github.com/jellyfin/jellyfin/blob/e1527857e931527f68b9705e248bd7bc83c62940/MediaBrowser.Model/Dlna/StreamBuilder.cs#L617-L618
`GetBitrateForDirectPlayCheck` doesn't use the user-defined bitrate limit.
https://github.com/jellyfin/jellyfin/blob/e1527857e931527f68b9705e248bd7bc83c62940/MediaBrowser.Model/Dlna/StreamBuilder.cs#L367-L370
That's why direct-playable video is not transcoded according to the user-defined bitrate limit.
